webapp/public/js/domjudge.js

```diff
@@ -1290,3 +1296,267 @@ function initScoreboardSubmissions() {
+            radios.change((e) => {
+                const diffMode = e.target.value;
+                f(diffMode);
+            });
+        },
+        'onDiffSelectChange': (f) => {
+            select.change((e) => {
+                const noDiff = e.target.value === "";
+                const submitId = parseInt(e.target.value);
+                f(submitId, noDiff);
+            });
+        }
+    };
+    editors[editorId] = editor;
+
+    const updateMode = (diffMode) => {
+        setDiffMode(diffMode);
+    };
+    updateMode(initialDiffMode);
+    editor.onDiffModeChange(updateMode);
+
+    const updateSelect = (submitId, noDiff) => {
+        radios.each((_, radio) => {
+            radio.disabled = noDiff;
+        });
+
+        const selected = select[0].options[select[0].selectedIndex];
+        if (selected && selected.dataset.tag) {
+            setDiffTag(selected.dataset.tag);
+        }
+
+        if (noDiff) {
+            diffTitle.style.display = 'none';
+        } else {
+            diffTitle.style.display = 'inline';
+            diffTag.innerText = selected.dataset.tag;
+            diffLink.href = selected.dataset.url;
+            diffLink.innerText = `s${submitId}`;
+        }
+    };
+    updateSelect(parseInt(select[0].value), select[0].value === "");
+    editor.onDiffSelectChange(updateSelect);
+}
+
+function ensureModel(models, submitId) {
+    if (submitId in models) {
+        const model = models[submitId];
+        if (!('model' in model)) {
+            model['model'] = monaco.editor.createModel(
+                model['source'],
+                undefined,
+                monaco.Uri.file("diff/" + submitId + "/" + model['filename'])
+            );
+        }
+        return model['model'];
+    }
+    const empty = monaco.editor.getModel(monaco.Uri.file("empty")) ?? monaco.editor.createModel("", undefined, monaco.Uri.file("empty"));
+    return empty;
+}
+
+function initDiffEditorTab(editorId, diffId, submissionId, models) {
+    const navItem = document.getElementById(`${diffId}-link`);
+    const isDeleted = !(submissionId in models);
+
+    const diffEditor = monaco.editor.createDiffEditor(
+        document.getElementById(diffId), {
+        scrollbar: {
+            alwaysConsumeMouseWheel: false,
+            vertical: 'auto',
+            horizontal: 'auto'
+        },
+        scrollBeyondLastLine: false,
+        automaticLayout: true,
+        readOnly: true,
+        theme: getCurrentEditorTheme(),
+    });
+
+    const updateMode = (diffMode) => {
+        diffEditor.updateOptions({
+            renderSideBySide: diffMode === 'side-by-side',
+        });
+    };
+    editors[editorId].onDiffModeChange(updateMode);
+
+    const renamedFrom = (oldName) => {
+        let renamed = navItem.querySelector('.renamed');
+        let arrow = navItem.querySelector('.fa-arrow-right');
+        if (oldName === undefined) {
+            if (renamed) {
+                navItem.removeChild(renamed);
```

Comment:
can you refactor `renamed` to `renamedTag` (or something better) I read it as a boolean first and wanted to comment about the state here.